minidump-processor/src/symbols.rs

```diff
@@ -350,3 +350,251 @@ pub fn simple_symbol_supplier(symbol_paths: Vec<PathBuf>) -> impl SymbolSupplier
 pub fn string_symbol_supplier(modules: HashMap<String, String>) -> impl SymbolSupplier {
     breakpad_symbols::StringSymbolSupplier::new(modules)
 }
+
+pub mod debuginfo {
+    use super::*;
+    use breakpad_symbols::SymbolFile;
+    use cachemap::CacheMap;
+    use futures_util::lock::Mutex;
+    use memmap2::Mmap;
+    use std::collections::HashMap;
+    use std::fs::File;
+    use std::path::Path;
+    use symbolic::{
+        cfi::CfiCache,
+        debuginfo::{self, Object},
+    };
+
+    /// A symbol provider which gets symbol information from the crashing binaries on the local
+    /// system.
+    #[derive(Default)]
+    pub struct DebugInfoSymbolProvider {
+        /// If a file fails to load for any reason, None is stored.
+        loaded: CacheMap<PathBuf, Lazy<Option<DebugInfo>>>,
+    }
+
+    #[derive(Default)]
+    struct Lazy<T>(Mutex<Option<T>>);
```

Comment:
FYI: you might consider using the `once_cell` crate's `Lazy` type here instead:
https://docs.rs/once_cell/latest/once_cell/sync/struct.Lazy.html

That crate is not very discoverable. Even though I've used it before it took me several minutes to remember what it was called.

Reply:
Looking a little closer, `once_cell`'s `Lazy` will be a bit suboptimal (though maybe in a premature optimization sort of way depending on the pattern of access) because:
1. It will implicitly require `F` to be static, so the path will need to be copied (not a huge deal), and
2. It will park a contentious thread rather than allowing other async tasks to run.

The second one is the real concern, as it will unnecessarily prevent progress. But it may not be a big deal. I could also submit a PR to `once_cell` to support an async `Lazy`.